### DISCUSSION TITLE
How to generate Demands array of DistanceMatrix array

### DISCUSSION THREAD

**Discussion by EngBassel** [google/or-tools#4360] [Routing (and legacy CP) questions]
Hello everyone,
in [Google OR-Tools CVRP](https://developers.google.com/optimization/routing/cvrp) shows an example of add vehicle capacities by using array of demands.
I need to know to can i generate Demands array of dynamic distance matrix array?

**Comment by sschnug** [comment]
It's totally unclear what your problem is and the closing of your (?) [StackOverflow Question](https://stackoverflow.com/questions/78918864/set-maximum-number-of-location-for-each-vehicle-in-google-or-tools) indicates that fact.

Please invest more than 10 seconds to formulate your question if you really want help!

**Reply by EngBassel** [reply]
as shown in google example [CVRP Example](https://developers.google.com/optimization/routing/cvrp#c_1) they used code:
`int demandCallbackIndex = routing.RegisterUnaryTransitCallback((long fromIndex) =>
                                                               {
                                                                   // Convert from routing variable Index to
                                                                   // demand NodeIndex.
                                                                   var fromNode =
                                                                       manager.IndexToNode(fromIndex);
                                                                   return data.Demands[fromNode];
                                                               });
routing.AddDimensionWithVehicleCapacity(demandCallbackIndex, 0, // null capacity slack
                                        data.VehicleCapacities, // vehicle maximum capacities
                                        true,                   // start cumul to zero
                                        "Capacity");`
for their distance matrix and array of demands.
in my program i have my own distance matrix but not demand array, so i need to know how to generate it.
Regards

**Reply by lperron** [reply]
What do you want to express with the demand array?

**Reply by EngBassel** [reply]
I need it to set maximum number of locations for each vehicle.

**Comment by lperron** [comment]
So demand is always 1, and max capacity of the dimension is the max number
of visits.

**Reply by EngBassel** [reply]
do you mean i have to create demand array containing only 1 value equal to 1?
i did that but it seems program enter to infinite loop! 
here is my full code:
```cs
public class DataModel
        {
            public long[,] DistanceMatrix = {
            { 0, 548, 776, 696, 582, 274, 502, 194, 308, 194, 536, 502, 388, 354, 468, 776, 662 },
            { 548, 0, 684, 308, 194, 502, 730, 354, 696, 742, 1084, 594, 480, 674, 1016, 868, 1210 },
            { 776, 684, 0, 992, 878, 502, 274, 810, 468, 742, 400, 1278, 1164, 1130, 788, 1552, 754 },
            { 696, 308, 992, 0, 114, 650, 878, 502, 844, 890, 1232, 514, 628, 822, 1164, 560, 1358 },
            { 582, 194, 878, 114, 0, 536, 764, 388, 730, 776, 1118, 400, 514, 708, 1050, 674, 1244 },
            { 274, 502, 502, 650, 536, 0, 228, 308, 194, 240, 582, 776, 662, 628, 514, 1050, 708 },
            { 502, 730, 274, 878, 764, 228, 0, 536, 194, 468, 354, 1004, 890, 856, 514, 1278, 480 },
            { 194, 354, 810, 502, 388, 308, 536, 0, 342, 388, 730, 468, 354, 320, 662, 742, 856 },
            { 308, 696, 468, 844, 730, 194, 194, 342, 0, 274, 388, 810, 696, 662, 320, 1084, 514 },
            { 194, 742, 742, 890, 776, 240, 468, 388, 274, 0, 342, 536, 422, 388, 274, 810, 468 },
            { 536, 1084, 400, 1232, 1118, 582, 354, 730, 388, 342, 0, 878, 764, 730, 388, 1152, 354 },
            { 502, 594, 1278, 514, 400, 776, 1004, 468, 810, 536, 878, 0, 114, 308, 650, 274, 844 },
            { 388, 480, 1164, 628, 514, 662, 890, 354, 696, 422, 764, 114, 0, 194, 536, 388, 730 },
            { 354, 674, 1130, 822, 708, 628, 856, 320, 662, 388, 730, 308, 194, 0, 342, 422, 536 },
            { 468, 1016, 788, 1164, 1050, 514, 514, 662, 320, 274, 388, 650, 536, 342, 0, 764, 194 },
            { 776, 868, 1552, 560, 674, 1050, 1278, 742, 1084, 810, 1152, 274, 388, 422, 764, 0, 798 },
            { 662, 1210, 754, 1358, 1244, 708, 480, 856, 514, 468, 354, 844, 730, 536, 194, 798, 0 }
        };
            public long Demands = 1;
            public long[] VehicleCapacities = { 4,4,4 };
            public int VehicleNumber = 3;
            public int Depot = 0;
};
public void CreateMatrix()
{
    DataModel data = new DataModel();
    // Create Routing Index Manager
    RoutingIndexManager manager = new RoutingIndexManager(data.DistanceMatrix.GetLength(0), data.VehicleNumber, data.Depot);
    // Create Routing Model.
    RoutingModel routing = new RoutingModel(manager);

    int transitCallbackIndex = routing.RegisterTransitCallback((long fromIndex, long toIndex) =>
    {
        // Convert from routing variable Index to
        // distance matrix NodeIndex.
        var fromNode = manager.IndexToNode(fromIndex);
        var toNode = manager.IndexToNode(toIndex);
        return data.DistanceMatrix[fromNode, toNode];
    });
    // Define cost of each arc.
    routing.SetArcCostEvaluatorOfAllVehicles(transitCallbackIndex);
    int demandCallbackIndex = routing.RegisterUnaryTransitCallback((long fromIndex) =>
    {
        // Convert from routing variable Index to
        // demand NodeIndex.
        var fromNode = manager.IndexToNode(fromIndex);
        return 1;
    });

    // Add Distance constraint.
    routing.AddDimensionWithVehicleCapacity(demandCallbackIndex, 0, data.VehicleCapacities,
                         true, // start cumul to zero
                         "capacity");
    // Setting first solution heuristic.
    RoutingSearchParameters searchParameters =
        operations_research_constraint_solver.DefaultRoutingSearchParameters();
    searchParameters.FirstSolutionStrategy = FirstSolutionStrategy.Types.Value.PathCheapestArc;

    // Solve the problem.
    Assignment solution = routing.SolveWithParameters(searchParameters);

    // Print solution on console.
    PrintSolution(data, routing, manager, solution);

}
```
